text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,629 @@
+
+In terms of the dependency graph, we would create one IR node
+representing the codegen unit. This would have the object code as an
+associated artifact. We would also have edges from each component of
+the codegen unit. As today, generic or inlined functions would not
+belong to any codegen unit, but rather would be instantiated anew into
+each codegen unit in which they are (transitively) referenced.
+
+There is an analogy here with C++, which naturally faces the same
+problems. In that setting, templates and inlineable functions are
+often placed into header files. Editing those header files naturally
+triggers more recompilation. The compiler could employ a similar
+strategy by replicating things that look like good candidates for
+inlining into each module; call graphs and profiling information may
+be a good input for such heuristics.
+
+<a id="testing"></a>
+## Testing strategy
+
+If we are not careful, incremental compilation has the potential to
+produce an infinite stream of irreproducible bug reports, so it's
+worth considering how we can best test this code.
+
+### Regression tests
+
+The first and most obvious piece of infrastructure is something for
+reliable regression testing. The plan is simply to have a series of
+sources and patches. The source will have each patch applied in
+sequence, rebuilding (incrementally) at each point. We can then check
+that (a) we only rebuilt what we expected to rebuild and (b) compare
+the result with the result of a fresh build from scratch.  This allows
+us to build up tests for specific scenarios or bug reports, but
+doesn't help with *finding* bugs in the first place.
+
+### Replaying crates.io versions and git history
+
+The next step is to search across crates.io for consecutive
+releases. For a given package, we can checkout version `X.Y` and then
+version `X.(Y+1)` and check that incrementally building from one to
+the other is successful and that all tests still yield the same
+results as before (pass or fail).
+
+A similar search can be performed across git history, where we
+identify pairs of consecutive commits. This has the advantage of being
+more fine-grained, but the disadvantage of being a MUCH larger search
+space.
+
+### Fuzzing
+
+The problem with replaying crates.io versions and even git commits is
+that they are probably much larger changes than the typical
+recompile. Another option is to use fuzzing, making "innocuous"
+changes that should trigger a recompile. Fuzzing is made easier here
+because we have an oracle -- that is, we can check that the results of
+recompiling incrementally match the results of compiling from scratch.
+It's also not necessary that the edits are valid Rust code, though we
+should test that too -- in particular, we want to test that the proper
+errors are reported when code is invalid, as well.  @nrc also
+suggested a clever hybrid, where we use git commits as a source for
+the fuzzer's edits, gradually building up the commit.
+
```

Comment:
These automated testing strategies (in addition to handwritten tests) seem like a very good idea to me.